feature/add-required-text

ineedsubstance opened the pull request:
My attempt to add `required to save/publish` text to the form labels.

I took a stab at this, but I'm sure it's not there yet. Specifically, I wasn't sure if I should be trying to render the label again or do a replace to add the new required text to the existing label.

---
Resolves https://github.com/punchcard-cms/punchcard/issues/404 (will still need to submit PR in Punchcard for CSS update)

`DCO 1.1 Signed-off-by: Nick Tilden <nick@ineedsubstance.com>`

Snugug commented:
As you can see from the failing test, you've got some issues off the bat; namely that `label` and `inp` aren't defined in the scope you're working in. Try starting there.

Snugug commented on lib/form/html.js:
```diff
@@ -169,6 +169,11 @@ const renderer = (type, errors) => {
       }
 
+      // Add required text to label
+      if (label) {
```
`label` is not defined

Snugug replied:
I'd also recommend you trying to run the tests locally to ensure they pass. This can be done by running `npm test` after installing the dependencies. You would have seen immediately that this fails.

You also need to include tests when adding new code, so please do so.

I also highly recommend you look at how input plugins are implemented (which this registers). Take, for instance, the HTML [number](https://github.com/punchcard-cms/input-plugin-number/blob/master/index.js#L32) and [quote](https://github.com/punchcard-cms/input-plugin-quote/blob/master/index.js#L59) input plugins and you'll see immediately that they both control their own `label` elements. Inputs can only have a single label, this information should be included as a text element with an [`aria-describeby`](http://webaim.org/techniques/forms/advanced#describedby) attribute on the input and ID on the description (will require some REGEX to actually dive in to the inputs to do this).

This needs to work for single input plugins as well as individual inputs in a multi-input input field (like quote)

ineedsubstance replied:
Thanks @Snugug I will try and address these issues.

Snugug commented on lib/form/html.js:
```diff
@@ -169,6 +169,11 @@ const renderer = (type, errors) => {
 
+      // Add required text to label
+      if (label) {
+        render += `<label for="${inp.id}" class="${required}">${label.id} <mark class="mark-${required}">required to ${required}</mark></label>`;
```
`inp` is not defined

scottnath commented:
Hi @ineedsubstance 

I didn't realize you wanted to tackle actual code! 

So, once you have the dev environment working and testing correctly...the actual place you want to make a change like this is in the [`addRequired` function](https://github.com/punchcard-cms/content-types/pull/111/files#diff-996a0736d729433497206723ee67a0c5L60). You'll need to dig into the regex to find the *end* of the label tag and tack on your `mark` there.

For testing, you'll want to look into `/tests/form.js` and `/tests/html.js` for getting your changes tested. For instance, in [this tested result](https://github.com/punchcard-cms/content-types/blob/adffae1b62700d46b7a4097ce36dffc0a6980e01/tests/form.js#L63) the system checks that the proper className is added to the `label`. You would want another check there to make sure that `mark` got properly added.

Hope this helps. Scott

ineedsubstance commented:
Thanks, @scottnath that helps. Yeah, I was hoping to do the styling and code if possible. My javascript is pretty weak, but I'm trying to figure it out so I can contribute where I can.

ineedsubstance commented:
@scottnath + @Snugug - Updated my approach based on Scott's recommendation and also got testing to work locally which helped immensely. Closer?

scottnath commented on lib/form/html.js:
```diff
@@ -137,6 +137,10 @@ const addRequired = (html, input, index) => {
+
+      // Add required text to label
+      const reLabel = new RegExp('</label>');
+      render = render.replace(reLabel, `<mark class="mark-${input.required}">required to ${input.required}</mark></label>`);
```
@ineedsubstance this is coming along great.
Now, you've got to account for both `inp.required` and `input.required`.
You could add a `required` variable up in the `if (inp.required !== undefined) {` conditional (line 93) to capture whether this is publish or save, then use that here.

ineedsubstance replied:
ok, let me try that. Thanks.

scottnath commented on tests/html.js:
```diff
@@ -10,7 +10,7 @@ test('Add Errors - Pass', t => {
     input: {
       description: 'I am the Bar Content Type Config textarea description',
-      html: '<label for=\'{{textarea.id}}\'>{{textarea.label}}</label><textarea id=\'{{textarea.id}}\' name=\'{{textarea.name}}\' placeholder=\'{{textarea.placeholder}}\' />{{textarea.value}}</textarea>',
+      html: '<label for=\'{{textarea.id}}\'>{{textarea.label}} <mark class=\'mark-{{textarea.required}}\'>required to {{textarea.required}}</mark></label><textarea id=\'{{textarea.id}}\' name=\'{{textarea.name}}\' placeholder=\'{{textarea.placeholder}}\' />{{textarea.value}}</textarea>',
```
Can you also add a test in the repeatable section below. We need to make sure this is working with repeatables too.

ineedsubstance replied:
Will do, thanks.

ineedsubstance commented:
@Snugug @scottnath - OK, let's try this again. 😁

Snugug commented on lib/form/html.js:
```diff
@@ -137,6 +138,10 @@ const addRequired = (html, input, index) => {
+
+      // Add required text to label
+      const reLabel = new RegExp('</label>');
+      render = render.replace(reLabel, `<mark class="mark-${required}">required to ${required}</mark></label>`);
```
I like the use of `mark` here

Snugug replied:
Mark's class should be the `level` you've defined above.

scottnath replied:
@Snugug do you mean this should be:
`required--publish`
or
`mark-required--publish`
?

Snugug replied:
It should match https://github.com/punchcard-cms/punchcard/pull/462

ineedsubstance replied:
Yeah, removing the `mark-`

Snugug commented on tests/html.js:
```diff
@@ -81,10 +81,10 @@ test('Add Errors Repeatable - Pass', t => {
 test('Add Required - Pass', t => {
   const param = {
-    html: '"<label for="91f79620-ba21-4a4a-a4c7-02f456129b0f">My Awesome Text Area</label><textarea id="91f79620-ba21-4a4a-a4c7-02f456129b0f" name="my-textarea--textarea" placeholder="Type something..." />test</textarea>"',
+    html: '"<label for="91f79620-ba21-4a4a-a4c7-02f456129b0f">My Awesome Text Area <mark class="mark-save">required to save</mark></label><textarea id="91f79620-ba21-4a4a-a4c7-02f456129b0f" name="my-textarea--textarea" placeholder="Type something..." />test</textarea>"',
```
This appears to be the only test that makes sure something's been rendered correctly. I'd like to see tests for a single input, an object, and repeatables, both `save` and `publish`, as well as when `require: true` is how it was created in the content type

scottnath commented on lib/form/html.js:
```diff
@@ -90,6 +90,7 @@ const addRequired = (html, input, index) => {
 
     if (input.required === 'save' || inp.required === 'save' || input.required === 'publish' || inp.required === 'publish') {
       let level = '';
+      const required = input.required && inp.required;
```
`&&` is not what you want, it would be `||` actually, this should be:
`const required = inp.required || input.required`
then you can get rid of the whole conditional:
```
const level = `required--${required}`
```

ineedsubstance replied:
Updated `const required = inp.required || input.required`

Tried removing the conditional statement below which caused the test to fail so I am leaving it in for now. 
```      
if (inp.required !== undefined) {
        level += `required--${inp.required}`;
      }
      else {
        level += `required--${input.required}`;
      }
```

scottnath replied:
@ineedsubstance did you add this line instead of the conditional?

```
const level = `required--${required}`;
```
if you removed the conditional without adding that line, then it would break the test, yes.

ineedsubstance replied:
I did not, will update.

ineedsubstance replied:
@scottnath - updated and tested locally but got the following error:

```
94:13  error  'level' is already defined  no-redeclare
```

Test passed after that.

So I removed this:
```
let level = '';
```

ineedsubstance commented:
@Snugug - in regards to this [comment](https://github.com/punchcard-cms/content-types/pull/111#discussion_r80543611), I added a test for a `repeatable` and an `object` and included required to publish in those. I wasn't sure if this was exactly what you were asking for, but here you go.

@scottnath - I [commented](https://github.com/punchcard-cms/content-types/pull/111#discussion_r80552710) on your [comment](https://github.com/punchcard-cms/content-types/pull/111#discussion_r80544885). Basically, it didn't pass the test, so I changed the first part but didn't remove the conditional statement.
